pkg/kubelet/container/runtime.go

```diff
@@ -375,6 +375,11 @@ type EnvVar struct {
 	Value string
 }
 
+type Annotation struct {
```

Comment:
Mind to explain a little bit who is expected to add these annotations and who is expect to consume them? Sorry for silly question if you've already clarified somewhere else.

Reply:
No worries :)

The annotations are expected to be added by device plugins only (i.e: here the NVIDIA device plugin).
CRI-O is expected to [consume these annotations](https://github.com/kubernetes-incubator/cri-o/blob/master/hooks.md).

Currently, the only NVIDIA supported way to add GPUs to containers is to run docker with the nvidia runtime as your default runtime (runc + a call to the nvidia OCI pre-start hook).

Adding annotations would allow us to use CRI-O without requesting users to setup the docker daemon.

Reply:
Thanks! It's useful for review. cc @Random-Liu for cri-containerd.

Reply:
@RenaudWasTaken I understand why you are doing this, and I think it's good to enable people experiment new ideas via Annotations. :)

But please note that annotation-based api/feature is usually considered unstable, because there is no api guarantee, and can be broken easily. It's good to experiment new things via annotation, but you may still want to eventually standardize the api and graduate from annotation.

As for the Device Plugin side, I think @jiayingz has more ideas about this.

But for the CRI side, please bring up the pre-start hook discussion in sig-node sometime, so that we can see whether we should support prestart hook in Kubernetes api, or support prestart hook in CRI, or at least standardize the prestart hook annotation format for alpha support.
/cc @yujuhong @kubernetes/sig-node-api-reviews

Reply:
That's great to hear!

With @3XX0 and @flx42 we wanted to wait after the device plugin beta to propose a standardized way to support pre-start hooks in CRI.

Reply:
+1 for @Random-Liu , pre-start hook is what I always want to discuss with you but forgotten. We currently injected extra information to container spec before CRI `CreateContainer` to enable NVLink for ML training. I really hope pre-start hook can help out. like: ./my-hook-binary <annotations>

Reply:
Prestart hook is certainly a feature that we'd consider to add in kubernetes. It was mentioned in the [initial CRI proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/container-runtime-interface-v1.md#container-lifecycle-hooks). Theoretically, CRI can support such hooks without any API change, but in reality, this needs to be called out and discussed since it's not something kubernetes already supports. +1 to what @Random-Liu suggested -- bringing up this discussion via a issue or in a sig-node meeting.

It's also worth mentioning that docker would not be able to support this out of the box AFAIK. I am not sure whether we can hack around it.

Reply:
@yujuhong We are not using docker but I also think it can be better implemented in other CRI shims like containerd